# Calculus and Python

## Motivation
I am starting the last course in a unified treatment of Calculus at a local community college. My learning process includes reading each chapter twice, creating notes, and solving end-of-section problems. End-of-Section problems increase in difficulty where having the solution manual is neccessary. But some problems don't have the solution or require more exploration. The goal of this repository is to promote an experimental mindset when applying Calculus and Python with the help of chatbots. 

## How to use
The comments in the code will contain the relevant Calculus concepts for easy reference and clarity. But please use other resources to learn the underlying Calculus concepts fully. 

## Notation and Style
Solutions are given in a way that makes sense to me and will closely follow the textbook and solution manual. Theorems, collolaries, formulas, etc. will consider textbook notation and style above all else.

## Citations
Swokowski, Earl W. *Calculus: The Classic Edition*. 7th ed., Brooks/Cole, 2012.

Swokowski, Earl W. *Student Solutions Manual for Calculus: The Classic Edition, Volume 2*. 7th ed., Brooks/Cole, 2012.

OpenAI. "ChatGPT". Accessed March 19, 2025. [https://www.openai.com/chatgpt].

# Install and Import

In [9]:
# Install the neccessary packages.
%pip install numpy sympy scipy pandas

# Import the neccessary libraries.
import numpy as np                      # For numerical operations and array handling.
import sympy as sp                      # For symbolic mathematics.
from scipy import integrate             # For numerical integration.
import pandas as pd                     # For data manipulation.

# Vectors and Surfaces

## Find Work Done

### Along y-axis from P1,P2
A constant force of magnitude 4 pounds has the same direction as the vector a = i + j + k. If distance is find the work done if the point of application moves along the y-axis from (0,2,0) to (0,-1,0).

In [10]:
# The magnitude of the constant force in pounds.
force_magnitude = 4

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
inital_point = np.array([0, 2, 0])
final_point = np.array([0, -1, 0])
displacement_vector = final_point - inital_point

# Define the direction vector. For this problem the direction vector a equals <1i + 1j + 1k>.
a = np.array([1, 1, 1]) 

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done, W is the dot product of the Force vector and the Displacement Vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in pounds (W):", W)

Force Vector (F): [2.30940108 2.30940108 2.30940108]
Displacement Vector (d): [ 0 -3  0]
Work Done in pounds (W): -6.9282032302755105


### Along a line from O to P
A constant force of magnitude 5 Newtons has the same direction as the positive z-axis. If distance is measured in meters, find the work done if the point of application moves along a line from the origin to the point P(1,2,3).

In [11]:
# The magnitude of the constant force in Newtons.
force_magnitude = 5

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
initial_point = np.array([0, 0, 0])
final_point = np.array([1, 2, 3])
displacement_vector = final_point - initial_point

# Define the direction vector. For this problem the direction vector a equals <0i + 0j + 1k>.
a = np.array([0, 0, 1]) 

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a 

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done, W is the dot product of the Force vector and the Displacement Vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in Newtons (W):", W)

Force Vector (F): [0. 0. 5.]
Displacement Vector (d): [1 2 3]
Work Done in Newtons (W): 15.0


### Angle with Horizontal
A person pulls a wagon along level ground by exerting a force of 20 pounds on a handle that makes an angle of 30 degrees with the horizontal. Find the work done in pulling the wagon 100 feet.

In [12]:
# The magnitude of the constant force in pounds.
force_magnitude = 20

# Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
initial_point = np.array([0, 0, 0])
final_point = np.array([100, 0, 0])
displacement_vector = final_point - initial_point

# Define the direction vector. The applied force makes a 30-degree angle with the horizontal.
# When force is applied at angle theta to the horizontal: x-component = cos(theta), y-component = sin(theta)
# When force is applied at angle theta to the vertical: y-component = cos(theta), x-component = sin(theta)

a = np.array([np.cos(np.radians(30)), np.sin(np.radians(30)), 0])

# The Unit vector in the direction of as defined by u = [1/norm(a)](a).
u = [1/np.linalg.norm(a)] * a 

# The Force vector equals 4a/norm(a).
force_vector = force_magnitude * u 

# Work done, W is the dot product of the Force vector and the Displacement Vector.
W = np.dot(force_vector, displacement_vector)

# Output the results
print("Force Vector (F):", force_vector)
print("Displacement Vector (d):", displacement_vector)
print("Work Done in Pounds (W):", W)

Force Vector (F): [17.32050808 10.          0.        ]
Displacement Vector (d): [100   0   0]
Work Done in Pounds (W): 1732.0508075688774


### Angle with x-axis and Incline (Not Done)
From the problem above, Force Vector is 17.32; Displacement Vector is <100,0,0>; Work done is 1732.05 . Find the work done if the wagon is pulled, with the same force, 100 feet up an incline that makes an angle of 30 degrees with the horizontal.

In [13]:
# # The magnitude of the constant force in pounds.
# # force_magnitude = 20

# force_magnitude = 17.32050808
# phi = 30
# displacement_along_incline = 100
# displacement_vector = np.array([displacement_along_incline, 0, 0])
# force_parallel = force_magnitude * np.cos(np.radians(phi))
# force_perpendicular = force_magnitude * np.sin(np.radians(phi))
# force_vector = np.array([force_parallel, force_perpendicular, 0])
# W = np.dot(force_vector, displacement_vector)

# # Define and Solve displacement vector from inital, final points. Note inital, final applications of force.
# initial_point = np.array([0, 0, 0])
# final_point = np.array([100, 0, 0])
# displacement_vector = final_point - initial_point

# # Define the direction vector. The applied force makes a 30-degree angle with the horizontal.
# # When force is applied at angle theta to the horizontal: x-component = cos(theta), y-component = sin(theta)
# # When force is applied at angle theta to the vertical: y-component = cos(theta), x-component = sin(theta)
# a = np.array([np.cos(np.radians(30)), np.sin(np.radians(30)), 0])

# # The Unit vector in the direction of as defined by u = [1/norm(a)](a).
# u = [1/np.linalg.norm(a)] * a 

# # The Force vector equals 4a/norm(a).
# force_vector = force_magnitude * u 

# # Work done, W is the dot product of the Force vector and the Displacement Vector.
# W = np.dot(force_vector, displacement_vector)

# # Output the results
# print("Force Vector (F):", force_vector)
# print("Displacement Vector (d):", displacement_vector)
# print("Work Done in Pounds (W):", W)

## Find the Angle between Vectors
Give a sequence A-B-C-D of four bonded atoms, the angle between the plane formed by A, B, and C as well as the plane formed by B, C, and D is called the torsion angle theta of the bond. This torsion angle is used to explain the stability of molecular structures. If segment BC is placed along the z-axis, how can theta be computed in terms of the components of vector BA and Bector CD?

In [14]:
# Step 1: Define coordinates for points A, B, C, and D.
# These represent the position of four bonded atoms in 3D space.
# Reminder: The problem states BC is aligned along the z-axis where
# zero's in B:<0, 0, 3> and C:<0, 0, 6>.
A = np.array([1, 2, 3])
B = np.array([0, 0, 3])
C = np.array([0, 0, 6])
D = np.array([2, -1, 6])

print("Coordinates:")
print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)
print()

# Step 2: Compute vectors BA, BC, and CD.
# These vectors represent the bonds between the atoms.
BA = A - B
BC = C - B
CD = D - C

print("Vectors:")
print("BA =", BA)
print("BC =", BC)
print("CD =", CD)
print()

# Step 3: Compute the normal vector n1 for Plane ABC using the determinant formula.
# The normal vector to a plane formed by two vectors (BA and BC) is given by:
# n1 = [a_y b_z - a_z b_y, a_z b_x - a_x b_z, a_x b_y - a_y b_x]
# Note: The vector product (or cross product) vector a cross vector b of vector a equals <a1, a2, a3> 
# and vector b equals <b1, b2, b3> is vector a cross vector b equals 
# matrix('a2 a3; b2 b3') - matrix('a1 a3; b1 b3') + matrix('a1 a2; b1 b2') can can be used instead.
n1_x = BA[1] * BC[2] - BA[2] * BC[1]
n1_y = BA[2] * BC[0] - BA[0] * BC[2]
n1_z = BA[0] * BC[1] - BA[1] * BC[0]
n1 = np.array([n1_x, n1_y, n1_z])

print("Normal vector to plane ABC (n1):")
print("n1_x =", n1_x, "n1_y =", n1_y, "n1_z =", n1_z)
print("n1 =", n1)
print()

# Step 4: Compute the normal vector n2 for Plane BCD using determinant formula.
# The normal vector to a plane formed by two vectors (BC and CD) is given by:
# n2 = [b_y c_z - b_z c_y, b_z c_x - b_x c_z, b_x c_y - b_y c_x].
n2_x = BC[1] * CD[2] - BC[2] * CD[1]
n2_y = BC[2] * CD[0] - BC[0] * CD[2]
n2_z = BC[0] * CD[1] - BC[1] * CD[0]
n2 = np.array([n2_x, n2_y, n2_z])

print("Normal vector to plane BCD (n2):")
print("n2_x =", n2_x, "n2_y =", n2_y, "n2_z =", n2_z)
print("n2 =", n2)
print()

# Step 5: Compute the torsion angle by using the dot product.
# If theta is the angle between nonzero vectors n1 and n2, then 
# cos(theta) = {DotP(n1,⋅n2) / [Norm(n1)][Norm(n2)]]}.
cos_theta = np.dot(n1, n2) / (np.linalg.norm(n1) * np.linalg.norm(n2))

print("Cosine of the torsion angle:")
print(cos_theta)
print()

# Step 6: Compute the angle in radians and convert to degrees.
# clip avoids numerical errors.
theta = np.arccos(np.clip(cos_theta, -1.0, 1.0))
theta_degrees = np.degrees(theta)

# Display results.
print("Torsion angle in radians:")
print(theta)
print()
print("Torsion angle in degrees:")
print(theta_degrees)

Coordinates:
A = [1 2 3]
B = [0 0 3]
C = [0 0 6]
D = [ 2 -1  6]

Vectors:
BA = [1 2 0]
BC = [0 0 3]
CD = [ 2 -1  0]

Normal vector to plane ABC (n1):
n1_x = 6 n1_y = -3 n1_z = 0
n1 = [ 6 -3  0]

Normal vector to plane BCD (n2):
n2_x = 3 n2_y = 6 n2_z = 0
n2 = [3 6 0]

Cosine of the torsion angle:
0.0

Torsion angle in radians:
1.5707963267948966

Torsion angle in degrees:
90.0
